drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,545 @@
+
+	return err;
+}
+
+static int adc_emul_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	return adc_emul_read_async(dev, sequence, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buf = data->buf;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	if (repeat_sampling) {
+		data->buf = data->repeat_buf;
+	}
+}
+
+/**
+ * @brief Convert input voltage of ADC @p chan to raw output value
+ *
+ * @param data Internal data of ADC emulator
+ * @param chan ADC channel to sample
+ * @param result Raw output value
+ *
+ * @return 0 on success
+ * @return -EINVAL if failed to get reference voltage or unknown input is
+ *         selected
+ * @return other error code returned by custom function
+ */
+static int adc_emul_get_chan_value(struct adc_emul_data *data,
+				   unsigned int chan,
+				   adc_emul_res_t *result)
+{
+	struct adc_emul_chan_cfg *chan_cfg = &data->chan_cfg[chan];
+	uint32_t input_mV;
+	uint32_t ref_v;
+	uint64_t temp; /* Temporary 64 bit value prevent overflows */
+	int err;
+
+	/* Get input voltage */
+	switch (chan_cfg->input) {
+	case ADC_EMUL_CONST_VALUE:
+		input_mV = chan_cfg->const_value;
+		break;
+
+	case ADC_EMUL_CUSTOM_FUNC:
+		err = chan_cfg->func(data->dev, chan, chan_cfg->func_data,
+				     &input_mV);
+		if (err) {
+			LOG_ERR("failed to read channel %d (err %d)",
+				chan, err);
+			return err;
+		}
+		break;
+
+	default:
+		LOG_ERR("unknown input source %d", chan_cfg->input);
+		return -EINVAL;
+	}
+
+	/* Get reference voltage and apply inverted gain */
+	ref_v = adc_emul_get_ref_voltage(data, chan_cfg->ref);
+	err = adc_gain_invert(chan_cfg->gain, &ref_v);
+	if (ref_v == 0 || err) {
+		LOG_ERR("failed to get ref voltage (channel %d)", chan);
+		return -EINVAL;
+	}
+
+	/* Calculate output value */
+	temp = (uint64_t)input_mV * data->res_mask / ref_v;
+
+	/* If output value is greater than resolution, it has to be trimmed */
+	if (temp > data->res_mask)
+		temp = data->res_mask;
+
+	*result = temp;
+
+	return 0;
+}
+
+/**
+ * @brief Main function of thread which is used to collect samples from
+ *        emulated ADC. When adc_context_start_sampling give semaphore,
+ *        for each requested channel value function is called. Returned
+ *        mV value is converted to output using reference voltage, gain
+ *        and requested resolution.
+ *
+ * @param data Internal data of ADC emulator
+ *
+ * @return This thread should not end
+ */
+static void adc_emul_acquisition_thread(struct adc_emul_data *data)
+{
+	int err;
+
+	while (true) {
+		k_sem_take(&data->sem, K_FOREVER);
+
+		err = 0;
+
+		while (data->channels) {
```

Comment:
Since this value is set from another thread, don't you need to use `volatile` access to it? What's the guarantee that this value is not cached in some way? Similarly for other values that can be changed after this thread is started (reference voltages, values for input channels),

Reply:
`channels`, `repeat_buf`, `buf` and `res_mask` fields of `struct adc_emul_data` are protected by `adc_context_lock()` and `adc_context_release()`. The `adc_emul_start_read()` which sets `res_mask` and `buf`, is called only once before `adc_context_start_read()`. `adc_context_start_sampling()` set `channels` and `repeat_buf`, then release acquisition thread. `adc_context_start_sampling()` and `adc_context_update_buffer_pointer()` will not be called before acquisition thread call `adc_context_on_sampling_done()`. Because of that I think there is no concurrent access to above `struct adc_emul_data` fields. Additionally acquisition thread take semaphore after all possible changes to these fields and I believe it is enough to clear cache.

As for reference voltages and channels configurations, I added `cfg_mtx` mutex which protect access to `ref_*` and `chan_cfg` fields of `struct adc_emul_data`.